AX: Collapse AccessibilityTableCell into AccessibilityRenderObject

hoffmanjoshua opened the pull request:
#### 8a02331ca45d6bed34882b7e7a78bcca50769af3
<pre>
AX: Collapse AccessibilityTableCell into AccessibilityRenderObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=298417">https://bugs.webkit.org/show_bug.cgi?id=298417</a>
<a href="https://rdar.apple.com/159890500">rdar://159890500</a>

Reviewed by Tyler Wilcock.

Collapses all of the logic from AccessibilityTableCell into AccessibilityRenderObject
and AccessibilityNodeObject as appropriate.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::hasCellOrRowRole const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::AXLogger::log):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::createFromNode):
(WebCore::AXObjectCache::handleRowspanChanged):
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::deferRowspanChange):
(WebCore::isAccessibilityTableCell): Deleted.
(WebCore::isAccessibilityARIAGridCell): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXObjectRareData.h:
(WebCore::AXObjectRareData::columnIndex const):
(WebCore::AXObjectRareData::setColumnIndex):
(WebCore::AXObjectRareData::axColIndexFromRow const):
(WebCore::AXObjectRareData::setAXColIndexFromRow):
(WebCore::AXObjectRareData::effectiveRowSpan const):
(WebCore::AXObjectRareData::incrementEffectiveRowSpan):
(WebCore::AXObjectRareData::resetEffectiveRowSpan):
* Source/WebCore/accessibility/AXTableHelpers.cpp:
(WebCore::AXTableHelpers::isTableCellElement):
* Source/WebCore/accessibility/AXTableHelpers.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRole):
(WebCore::AccessibilityNodeObject::computeIsIgnored const):
(WebCore::AccessibilityNodeObject::rowHeaders):
(WebCore::AccessibilityNodeObject::computeCellSlots):
(WebCore::AccessibilityNodeObject::parentTable const):
(WebCore::AccessibilityNodeObject::setRowIndex):
(WebCore::AccessibilityNodeObject::axColumnIndex const):
(WebCore::AccessibilityNodeObject::axRowIndex const):
(WebCore::AccessibilityNodeObject::axRowIndexText const):
(WebCore::AccessibilityNodeObject::isTableCell const):
(WebCore::AccessibilityNodeObject::isARIAGridCell const):
(WebCore::AccessibilityNodeObject::isExposedTableCell const):
(WebCore::AccessibilityNodeObject::parentTableIfTableCell const):
(WebCore::AccessibilityNodeObject::isTableHeaderCell const):
(WebCore::AccessibilityNodeObject::isColumnHeader const):
(WebCore::AccessibilityNodeObject::isRowHeader const):
(WebCore::AccessibilityNodeObject::rowIndexRange const):
(WebCore::AccessibilityNodeObject::columnIndexRange const):
(WebCore::AccessibilityNodeObject::axColumnIndexText const):
(WebCore::AccessibilityNodeObject::colSpan const):
(WebCore::AccessibilityNodeObject::rowSpan const):
(WebCore::AccessibilityNodeObject::incrementEffectiveRowSpan):
(WebCore::AccessibilityNodeObject::resetEffectiveRowSpan):
(WebCore::AccessibilityNodeObject::setAXColIndexFromRow):
(WebCore::AccessibilityNodeObject::setColumnIndex):
(WebCore::AccessibilityNodeObject::parentRow const):
(WebCore::AccessibilityNodeObject::axColumnSpan const):
(WebCore::AccessibilityNodeObject::axRowSpan const):
(WebCore::AccessibilityNodeObject::ensureIndexesUpToDate const):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::readOnlyValue const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::titleUIElement const):
(WebCore::AccessibilityRenderObject::computeIsIgnored const):
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
(WebCore::AccessibilityRenderObject::addChildren):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/accessibility/AccessibilityTableCell.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityTableCell.h: Removed.
* Source/WebCore/accessibility/AccessibilityTree.cpp:
* Source/WebCore/accessibility/atspi/AccessibilityObjectAtspi.cpp:
(WebCore::AccessibilityObjectAtspi::attributes const):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper tableCellParent]):
(-[WebAccessibilityObjectWrapper accessibilityHeaderElements]):
(-[WebAccessibilityObjectWrapper accessibilityARIARowIndex]):
(-[WebAccessibilityObjectWrapper accessibilityARIAColumnIndex]):
(-[WebAccessibilityObjectWrapper accessibilityRowIndexDescription]):
(-[WebAccessibilityObjectWrapper accessibilityColumnIndexDescription]):
(-[WebAccessibilityObjectWrapper accessibilityColumnRange]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::parentTableIfTableCell const):
(WebCore::AXIsolatedObject::parentTable const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::objectChangedIgnoredState):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:

Canonical link: <a href="https://commits.webkit.org/299635@main">https://commits.webkit.org/299635@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/990cd650d7a77f4cbb9878e6938fdff3af078f59

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119645 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39339 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29991 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125932 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71719 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121521 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40035 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47917 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90876 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/60178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122596 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31964 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107282 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71399 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31000 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25387 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69577 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101432 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25579 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128893 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46567 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35277 "Found 1 new test failure: http/tests/cache/disk-cache/disk-cache-302-status-code.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99477 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46932 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103475 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99318 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25231 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44746 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22766 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43131 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46429 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52135 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45895 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49244 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47581 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXTableHelpers.cpp:
```diff
@@ -104,6 +104,22 @@ bool isTableRowElement(Element& element)
 
+bool isTableCellElement(Element& element)
+{
+    if (is<HTMLTableCellElement>(element))
+        return true;
```
What if it has a non-cell role?

hoffmanjoshua replied:
This matches our existing behavior, where we create `AccessibilityTableCell` objects for anything that matches this criteria. But, we should be more specific about this. I updated this to consider if something has a non-cell aria role, and no tests fail.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -411,7 +411,15 @@ AccessibilityRole AccessibilityNodeObject::determineAccessibilityRole()
         return AccessibilityRole::Row;
 
-    return determineAccessibilityRoleFromNode();
+    AccessibilityRole roleFromNode = determineAccessibilityRoleFromNode();
```
`auto` seems appropriate here considering the variable name and method initializing the variable have "role" in it?

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2521,6 +2627,262 @@ AXCoreObject* AccessibilityNodeObject::parentTableIfExposedTableRow() const
+        return false;
+
+    RefPtr parentTable = this->parentTable();
+    bool result = parentTable && parentTable->isExposableTable();
+    return result;
```
Why have a separate bool local variable?

hoffmanjoshua replied:
Fixed, this should have been one line but this was for testing

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -885,6 +893,10 @@ bool AccessibilityNodeObject::computeIsIgnored() const
         return true;
     }
+
+    if (isTableCell())
+        return !isExposedTableCell();
```
It's a bit unfortunate that we call `isTableCell()` twice here, but probably not a huge deal. One idea would be to have a method called `isIgnorableTableCell()` that returns false for `isTableCell()`, or true for no exposable ancestor table exists. What do you think?

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1985,11 +1997,33 @@ bool AccessibilityNodeObject::isDataTable() const
+            return headers;
+
+        auto rowRange = rowIndexRange();
+        auto colRange = columnIndexRange();
```
I think WebKit prefers to explicitly spell out most names, e.g. `columnRange` rather than `colRange`

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1985,11 +1997,33 @@ bool AccessibilityNodeObject::isDataTable() const
+
+        for (unsigned column = 0; column < colRange.first; column++) {
+            RefPtr tableCell = parent->cellForColumnAndRow(column, rowRange.first);
+            if (!tableCell || tableCell == this || headers.contains(Ref { *tableCell }))
```
`headers.contains(Ref { *tableCell })`

A more efficient implementation is:

```cpp
headers.containsIf([] (const auto& header) {
    return header == tableCell.get();
})
```

This avoids ref-count churn

hoffmanjoshua replied:
I had to modify this to work with the Ref from the `header`, but now we don't create a new Ref each time.

twilco commented on Source/WebCore/accessibility/AXTableHelpers.cpp:
```diff
@@ -104,6 +104,23 @@ bool isTableRowElement(Element& element)
 
+bool isTableCellElement(Element& element)
+{
+    bool hasCellRole = hasCellARIARole(element);
+    if (is<HTMLTableCellElement>(element) && (hasCellRole || hasRole(element, nullAtom())))
+        return true;
+
+    if (hasCellRole)
+        return true;
```
If we're going to unconditionally return true `if (hasCellRole)`, should that come first? That saves an `is<HTMLTableCellElement>(element)` check

hoffmanjoshua replied:
Yeah, that's a good call. We can actually make this a little cleaner if we reorganize it that way, preventing the need for an extra bool:
```cpp
    if(hasCellARIARole(element))
        return true
    if (is<HTMLTableCellElement>(element) && hasRole(element, nullAtom()))
        return true;
```

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2521,6 +2629,261 @@ AXCoreObject* AccessibilityNodeObject::parentTableIfExposedTableRow() const
+    if (!element)
+        return false;
+
+    if (isARIAGridCell())
+        return true;
```
Doesn't `AXTableHelpers::isTableCellElement` already check this?

hoffmanjoshua replied:
Nice catch! I used to have it split out, but since than I had included this check in `isTableCellElement`. I've removed it

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2521,6 +2629,261 @@ AXCoreObject* AccessibilityNodeObject::parentTableIfExposedTableRow() const
+    if (!isTableCell())
+        return;
+
+    if (!hasRareData() && !index)
+        return;
```
Could we leave a comment explaining that this is an optimization (I presume I'm understanding that correctly?) to avoid creating a rare data when we don't need to? Or am I reading this wrong? This took me a little while to wrap my head around

twilco replied:
I also kind of wonder whether something that's a table cell is going to always have rare data created for it anyways through some other field, making this check a bit unnecessary? Or do you think it's reasonable that we'd have a number of cells benefit from this in common table layouts?

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.h:
```diff
@@ -131,6 +131,29 @@ class AccessibilityNodeObject : public AccessibilityObject {
+    AccessibilityObject* parentTableIfTableCell() const final;
+    bool isTableHeaderCell() const;
+    bool isColumnHeader() const final;
+    bool isRowHeader() const override;
```
Can this be final?

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1035,6 +1034,44 @@ AccessibilityObject* AccessibilityRenderObject::titleUIElement() const
+        if (!col)
+            return nullptr;
+
+        int row = renderCell->rowIndex();
```
I know you didn't write this, but should we move this to after the `CheckedPtr section`, closer to where it's actually used?

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3165fb0de2ca12ed7f64acd4a5a8aebebbf195ef. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50323

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/990cd650d7a77f4cbb9878e6938fdff3af078f59. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50323

hoffmanjoshua committed:
AX: Collapse AccessibilityTableCell into AccessibilityRenderObject
https://bugs.webkit.org/show_bug.cgi?id=298417
rdar://159890500

Reviewed by Tyler Wilcock.

Collapses all of the logic from AccessibilityTableCell into AccessibilityRenderObject
and AccessibilityNodeObject as appropriate.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::hasCellOrRowRole const):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::AXLogger::log):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::createFromNode):
(WebCore::AXObjectCache::handleRowspanChanged):
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::deferRowspanChange):
(WebCore::isAccessibilityTableCell): Deleted.
(WebCore::isAccessibilityARIAGridCell): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXObjectRareData.h:
(WebCore::AXObjectRareData::columnIndex const):
(WebCore::AXObjectRareData::setColumnIndex):
(WebCore::AXObjectRareData::axColIndexFromRow const):
(WebCore::AXObjectRareData::setAXColIndexFromRow):
(WebCore::AXObjectRareData::effectiveRowSpan const):
(WebCore::AXObjectRareData::incrementEffectiveRowSpan):
(WebCore::AXObjectRareData::resetEffectiveRowSpan):
* Source/WebCore/accessibility/AXTableHelpers.cpp:
(WebCore::AXTableHelpers::isTableCellElement):
* Source/WebCore/accessibility/AXTableHelpers.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRole):
(WebCore::AccessibilityNodeObject::computeIsIgnored const):
(WebCore::AccessibilityNodeObject::rowHeaders):
(WebCore::AccessibilityNodeObject::computeCellSlots):
(WebCore::AccessibilityNodeObject::parentTable const):
(WebCore::AccessibilityNodeObject::setRowIndex):
(WebCore::AccessibilityNodeObject::axColumnIndex const):
(WebCore::AccessibilityNodeObject::axRowIndex const):
(WebCore::AccessibilityNodeObject::axRowIndexText const):
(WebCore::AccessibilityNodeObject::isTableCell const):
(WebCore::AccessibilityNodeObject::isARIAGridCell const):
(WebCore::AccessibilityNodeObject::isExposedTableCell const):
(WebCore::AccessibilityNodeObject::parentTableIfTableCell const):
(WebCore::AccessibilityNodeObject::isTableHeaderCell const):
(WebCore::AccessibilityNodeObject::isColumnHeader const):
(WebCore::AccessibilityNodeObject::isRowHeader const):
(WebCore::AccessibilityNodeObject::rowIndexRange const):
(WebCore::AccessibilityNodeObject::columnIndexRange const):
(WebCore::AccessibilityNodeObject::axColumnIndexText const):
(WebCore::AccessibilityNodeObject::colSpan const):
(WebCore::AccessibilityNodeObject::rowSpan const):
(WebCore::AccessibilityNodeObject::incrementEffectiveRowSpan):
(WebCore::AccessibilityNodeObject::resetEffectiveRowSpan):
(WebCore::AccessibilityNodeObject::setAXColIndexFromRow):
(WebCore::AccessibilityNodeObject::setColumnIndex):
(WebCore::AccessibilityNodeObject::parentRow const):
(WebCore::AccessibilityNodeObject::axColumnSpan const):
(WebCore::AccessibilityNodeObject::axRowSpan const):
(WebCore::AccessibilityNodeObject::ensureIndexesUpToDate const):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::readOnlyValue const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::titleUIElement const):
(WebCore::AccessibilityRenderObject::computeIsIgnored const):
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
(WebCore::AccessibilityRenderObject::addChildren):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/accessibility/AccessibilityTableCell.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityTableCell.h: Removed.
* Source/WebCore/accessibility/AccessibilityTree.cpp:
* Source/WebCore/accessibility/atspi/AccessibilityObjectAtspi.cpp:
(WebCore::AccessibilityObjectAtspi::attributes const):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper tableCellParent]):
(-[WebAccessibilityObjectWrapper accessibilityHeaderElements]):
(-[WebAccessibilityObjectWrapper accessibilityARIARowIndex]):
(-[WebAccessibilityObjectWrapper accessibilityARIAColumnIndex]):
(-[WebAccessibilityObjectWrapper accessibilityRowIndexDescription]):
(-[WebAccessibilityObjectWrapper accessibilityColumnIndexDescription]):
(-[WebAccessibilityObjectWrapper accessibilityColumnRange]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::parentTableIfTableCell const):
(WebCore::AXIsolatedObject::parentTable const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::objectChangedIgnoredState):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:

Canonical link: https://commits.webkit.org/299635@main
